File: python/samples/agents/function_tools_agent/src/agent.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import os.path as path
import re
import sys
import json
import traceback
from datetime import datetime, timezone
from random import randint
from dotenv import load_dotenv
from pathlib import Path

from os import environ
from openai import AsyncOpenAI
from microsoft_agents.hosting.aiohttp import CloudAdapter
from microsoft_agents.hosting.core import (
    Authorization,
    AgentApplication,
    TurnState,
    TurnContext,
    MemoryStorage,
)
from microsoft_agents.authentication.msal import MsalConnectionManager
from microsoft_agents.activity import load_configuration_from_env

# Load .env from repository root
env_path = Path(__file__).resolve().parent.parent.parent.parent.parent.parent / ".env"
load_dotenv(env_path)

agents_sdk_config = load_configuration_from_env(environ)

STORAGE = MemoryStorage()

# Simple anonymous mode - no authentication required
try:
    CONNECTION_MANAGER = MsalConnectionManager(**agents_sdk_config)
    ADAPTER = CloudAdapter(connection_manager=CONNECTION_MANAGER)
    AUTHORIZATION = Authorization(STORAGE, CONNECTION_MANAGER, **agents_sdk_config)
except (ValueError, Exception):
    # If auth config is missing, create without authentication for local testing
    CONNECTION_MANAGER = None
    ADAPTER = CloudAdapter()
    AUTHORIZATION = None

AGENT_APP = AgentApplication[TurnState](
    storage=STORAGE,
    adapter=ADAPTER,
    authorization=AUTHORIZATION
)

# OpenAI client - can be overridden for testing
_openai_client = None
foundry_model = environ.get("FOUNDRY_MODEL_DEPLOYMENT", "gpt-5-nano")


def create_openai_client(injected_client=None):
    """Create OpenAI client for production or testing.

    Args:
        injected_client: Optional pre-configured client (for testing)

    Returns:
        AsyncOpenAI client instance
    """
    if injected_client is not None:
        return injected_client

    # Production path - create from environment variables
    foundry_endpoint = environ.get("FOUNDRY_ENDPOINT")
    foundry_api_key = environ.get("FOUNDRY_API_KEY")

    if not foundry_endpoint or not foundry_api_key:
        raise ValueError("FOUNDRY_ENDPOINT and FOUNDRY_API_KEY environment variables are required")

    return AsyncOpenAI(
        api_key=foundry_api_key,
        base_url=f"{foundry_endpoint}/openai/v1"
    )


def set_openai_client(client):
    """Override OpenAI client for testing.

    Args:
        client: AsyncOpenAI client or mock/recording wrapper
    """
    global _openai_client
    _openai_client = client


def get_openai_client():
    """Get OpenAI client (lazy initialization).

    Returns:
        AsyncOpenAI client instance (real or injected)
    """
    global _openai_client
    if _openai_client is None:
        _openai_client = create_openai_client()
    return _openai_client


# Conversation history storage (in-memory for demo)
conversation_history = {}


# Function Tools
def get_weather(location: str) -> str:
    """Get the weather for a given location."""
    # Use deterministic seed for testing (based on location)
    import random
    seed_value = sum(ord(c) for c in location) + 42
    random.seed(seed_value)

    conditions = ["sunny", "cloudy", "rainy", "partly cloudy", "stormy"]
    condition = conditions[random.randint(0, len(conditions) - 1)]
    temperature = random.randint(10, 35)
    return f"🌤️ The weather in {location} is {condition} with a temperature of {temperature}°C."


def get_time() -> str:
    """Get the current UTC time."""
    now = datetime.now(timezone.utc)
    return f"🕐 The current UTC time is {now.strftime('%Y-%m-%d %H:%M:%S')}."


def extract_location(message: str) -> str:
    """Extract location from user message."""
    patterns = [" in ", " at ", " for "]
    for pattern in patterns:
        index = message.lower().find(pattern)
        if index >= 0:
            location_start = index + len(pattern)
            location_part = message[location_start:].strip(' ?!.')
            if location_part:
                words = location_part.split(' ')
                return words[0].capitalize()
    return "your location"


@AGENT_APP.conversation_update("membersAdded")
async def on_members_added(context: TurnContext, _state: TurnState):
    await context.send_activity(
        "Hello! I'm a Function Tools Agent. "
        "I can help you with weather and time information. "
        "Try asking: 'What's the weather in Seattle?' or 'What time is it?'"
    )
    return True


@AGENT_APP.activity("message")
async def on_message(context: TurnContext, _state: TurnState):
    # Extract role from channelData (default to "user" if not present)
    role = "user"
    if hasattr(context.activity, "channel_data") and isinstance(context.activity.channel_data, dict):
        role = context.activity.channel_data.get("role", "user")

    # Only respond to user messages
    if role != "user":
        return

    user_message = context.activity.text if context.activity.text else ""
    conversation_id = context.activity.conversation.id

    # Initialize conversation history if needed
    if conversation_id not in conversation_history:
        conversation_history[conversation_id] = [
            {
                "role": "system",
                "content": "You are a helpful assistant that can check the weather and tell the time. Use the available functions to help users."
            }
        ]

    # Add user message to history
    conversation_history[conversation_id].append({
        "role": "user",
        "content": user_message
    })

    # Define available functions
    tools = [
        {
            "type": "function",
            "function": {
                "name": "get_weather",
                "description": "Get the weather for a given location.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "location": {
                            "type": "string",
                            "description": "The location to get the weather for."
                        }
                    },
                    "required": ["location"]
                }
            }
        },
        {
            "type": "function",
            "function": {
                "name": "get_time",
                "description": "Get the current UTC time.",
                "parameters": {
                    "type": "object",
                    "properties": {}
                }
            }
        }
    ]

    # Call LLM with function calling in a loop
    response_text = ""
    max_iterations = 5
    iteration = 0

    while iteration < max_iterations:
        iteration += 1

        # Get completion from LLM (using injected or real client)
        client = get_openai_client()
        completion = await client.chat.completions.create(
            model=foundry_model,
            messages=conversation_history[conversation_id],
            tools=tools,
            tool_choice="auto",
            temperature=0.0,  # Deterministic for testing
            seed=42  # Additional determinism
        )

        choice = completion.choices[0]
        message = choice.message

        # Check if the model wants to call functions
        if choice.finish_reason == "tool_calls" and message.tool_calls:
            # Add assistant message with tool calls to history
            conversation_history[conversation_id].append({
                "role": "assistant",
                "content": message.content,
                "tool_calls": [
                    {
                        "id": tc.id,
                        "type": tc.type,
                        "function": {
                            "name": tc.function.name,
                            "arguments": tc.function.arguments
                        }
                    }
                    for tc in message.tool_calls
                ]
            })

            # Execute each tool call
            for tool_call in message.tool_calls:
                function_name = tool_call.function.name
                function_args = json.loads(tool_call.function.arguments)

                if function_name == "get_weather":
                    location = function_args.get("location", "unknown")
                    function_result = get_weather(location)
                elif function_name == "get_time":
                    function_result = get_time()
                else:
                    function_result = "Unknown function"

                # Add function result to conversation history
                conversation_history[conversation_id].append({
                    "role": "tool",
                    "tool_call_id": tool_call.id,
                    "content": function_result
                })
        else:
            # Model provided a final response
            response_text = message.content or "I apologize, but I wasn't able to complete your request."
            conversation_history[conversation_id].append({
                "role": "assistant",
                "content": response_text
            })
            break

    if not response_text:
        response_text = "I apologize, but I wasn't able to complete your request."

    await context.send_activity(response_text)


@AGENT_APP.error
async def on_error(context: TurnContext, error: Exception):
    # This check writes out errors to console log .vs. app insights.
    # NOTE: In production environment, you should consider logging this to Azure
    #       application insights.
    print(f"\n [on_turn_error] unhandled error: {error}", file=sys.stderr)
    traceback.print_exc()

    # Send a message to the user
    await context.send_activity("The bot encountered an error or bug.")

File: typescript/samples/basic-m365/src/agent.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { TurnState, MemoryStorage, TurnContext, AgentApplication, AttachmentDownloader }
  from '@microsoft/agents-hosting'
import { version } from '@microsoft/agents-hosting/package.json'
import { ActivityTypes } from '@microsoft/agents-activity'
import OpenAI from 'openai'
import * as fs from 'fs'
import * as path from 'path'

interface ConversationState {
  messages: OpenAI.Chat.ChatCompletionMessageParam[];
}
type ApplicationTurnState = TurnState<ConversationState>

const downloader = new AttachmentDownloader()

// 🔧 FIX: Track storage access for cleanup to prevent memory leaks
const storageAccessTimes = new Map<string, number>()
const STORAGE_MAX_AGE_MS = 60 * 60 * 1000 // 1 hour
let lastCleanup = Date.now()

function cleanupOldStorage(storage: MemoryStorage): void {
  const now = Date.now()

  // Only cleanup every 5 minutes
  if (now - lastCleanup < 5 * 60 * 1000) {
    return
  }

  lastCleanup = now
  const cutoff = now - STORAGE_MAX_AGE_MS

  // Find keys to delete
  const keysToDelete: string[] = []
  for (const [key, accessTime] of storageAccessTimes.entries()) {
    if (accessTime < cutoff) {
      keysToDelete.push(key)
    }
  }

  // Delete old keys
  if (keysToDelete.length > 0) {
    storage.delete(keysToDelete).then(() => {
      keysToDelete.forEach(key => storageAccessTimes.delete(key))
      console.log(`Cleaned up ${keysToDelete.length} old conversation states`)
    }).catch(err => {
      console.error('Error during storage cleanup:', err)
    })
  }
}

// LLM Client setup
let openaiClient: OpenAI | null = null
let model: string = 'gpt-4'
let useRecordings: boolean = false

function initLLM() {
  // Check if we should use LLM recordings (test mode)
  useRecordings = process.env.USE_LLM_RECORDINGS?.toLowerCase() === 'true'

  if (useRecordings) {
    // Test mode: Use recorded LLM responses
    model = 'gpt-5-nano'  // Default model for recordings
    console.log('▶️  LLM Playback enabled (using recordings)')
    console.log('   Using recorded LLM responses (test mode)')
  } else {
    // Generation mode: Use real LLM
    const endpoint = process.env.FOUNDRY_ENDPOINT
    const apiKey = process.env.FOUNDRY_API_KEY

    if (!endpoint || !apiKey) {
      console.log('⚠️  FOUNDRY_ENDPOINT or FOUNDRY_API_KEY not set. LLM features disabled.')
      console.log('   Set these environment variables to enable LLM functionality.')
      return
    }

    model = process.env.FOUNDRY_MODEL_DEPLOYMENT || 'gpt-4'

    // Create OpenAI client for Foundry
    openaiClient = new OpenAI({
      apiKey: apiKey,
      baseURL: `${endpoint}/openai/v1/`
    })

    // Check if LLM recording is enabled
    const recordLlm = process.env.RECORD_LLM?.toLowerCase() === 'true'
    if (recordLlm) {
      console.log('📹 LLM Recording enabled (not implemented yet)')
    }
  }
}

// Initialize on module load
initLLM()

// Define storage and application
const storage = new MemoryStorage()
export const agentApp = new AgentApplication<ApplicationTurnState>({
  storage,
  fileDownloaders: [downloader]
})

// Function tools
async function getWeatherAsync(location: string): Promise<string> {
  // Simulate async API call
  await new Promise(resolve => setTimeout(resolve, 100))

  const conditions = ['sunny', 'cloudy', 'rainy', 'partly cloudy', 'stormy']
  const condition = conditions[Math.floor(Math.random() * conditions.length)]
  const temperature = Math.floor(Math.random() * 25) + 10

  return `🌤️ The weather in ${location} is ${condition} with a temperature of ${temperature}°C.`
}

function getCurrentTime(): string {
  const now = new Date()
  return `🕐 The current UTC time is ${now.toISOString().replace('T', ' ').substring(0, 19)}.`
}

// Convert OpenAI chat message to Agent Protocol message format
function convertChatMessageToAgentProtocol(chatMessage: OpenAI.Chat.ChatCompletionMessageParam): any {
  const message: any = {}

  // Determine role
  const role = chatMessage.role
  if (role === 'tool') {
    message.role = 'tool'
  } else {
    message.role = 'assistant'
  }

  // Convert contents
  const contents: any[] = []

  if (role === 'assistant' && 'tool_calls' in chatMessage && chatMessage.tool_calls) {
    // Tool call message
    for (const toolCall of chatMessage.tool_calls) {
      contents.push({
        kind: 'functionCall',
        callId: toolCall.id,
        name: toolCall.function.name,
        arguments: toolCall.function.arguments
      })
    }
  } else if (role === 'tool') {
    // Tool result
    const toolMsg = chatMessage as OpenAI.Chat.ChatCompletionToolMessageParam
    contents.push({
      kind: 'functionResult',
      callId: toolMsg.tool_call_id,
      result: toolMsg.content
    })
  } else if (role === 'assistant' && 'content' in chatMessage) {
    // Text response
    const text = chatMessage.content || ''
    contents.push({
      kind: 'text',
      text: text
    })
  } else {
    // Fallback: treat as text
    const text = ('content' in chatMessage) ? chatMessage.content : ''
    contents.push({
      kind: 'text',
      text: text || ''
    })
  }

  message.contents = contents
  return message
}

agentApp.onConversationUpdate('membersAdded', async (context: TurnContext, state: ApplicationTurnState) => {
  await context.sendActivity(
    "Hello! I'm a Basic M365 Agent. I can help you with weather and time information. " +
    "Try asking: 'What's the weather in Seattle?' or 'What time is it?'"
  )
})

// Listen for ANY message to be received
agentApp.onActivity(ActivityTypes.Message, async (context: TurnContext, state: ApplicationTurnState) => {
  // 🔧 Periodically clean up old storage to prevent memory leaks
  cleanupOldStorage(storage)

  // Track this conversation's access time
  if (context.activity.conversation?.id) {
    storageAccessTimes.set(context.activity.conversation.id, Date.now())
  }

  // Extract role from channelData (default to "user" if not present)
  const role = (context.activity.channelData as any)?.role ?? 'user'

  // Only respond to user messages
  if (role !== 'user') {
    return
  }

  const userMessage = context.activity.text || ''

  // If LLM is not configured, just echo
  if (!openaiClient && !useRecordings) {
    await context.sendActivity(
      `Echo: ${userMessage}\n\n(Note: LLM not configured. Set FOUNDRY_ENDPOINT and FOUNDRY_API_KEY to enable LLM features.)`
    )
    return
  }

  // Initialize conversation history if needed
  if (!state.conversation.messages) {
    state.conversation.messages = [
      {
        role: 'system',
        content: 'You are a helpful assistant that can check the weather and tell the time. Use the available functions to help users.'
      }
    ]
  }

  // Add user message to history
  state.conversation.messages.push({
    role: 'user',
    content: userMessage
  })

  // Track the starting index to capture new messages generated during this turn
  const startingHistoryCount = state.conversation.messages.length

  // Define available functions
  const tools: OpenAI.Chat.ChatCompletionTool[] = [
    {
      type: 'function',
      function: {
        name: 'GetWeatherAsync',
        description: 'Get the weather for a given location.',
        parameters: {
          type: 'object',
          properties: {
            location: {
              type: 'string',
              description: 'The location to get the weather for.'
            }
          },
          required: ['location']
        }
      }
    },
    {
      type: 'function',
      function: {
        name: 'GetCurrentTime',
        description: 'Get the current UTC time.',
        parameters: {
          type: 'object',
          properties: {}
        }
      }
    }
  ]

  // Call LLM with function calling in a loop
  let response = ''
  const maxIterations = 5  // Prevent infinite loops
  let iteration = 0

  while (iteration < maxIterations) {
    iteration++

    try {
      let completion: OpenAI.Chat.ChatCompletion

      if (useRecordings) {
        // Test mode: Return a simple mock response
        completion = {
          id: 'mock',
          object: 'chat.completion',
          created: Date.now(),
          model: model,
          choices: [{
            index: 0,
            message: {
              role: 'assistant',
              content: 'I can help you with weather and time information!',
              refusal: null
            },
            finish_reason: 'stop',
            logprobs: null
          }],
          usage: { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0 }
        }
      } else if (openaiClient) {
        // Generation mode: Use real LLM
        completion = await openaiClient.chat.completions.create({
          model: model,
          messages: state.conversation.messages,
          tools: tools
        })
      } else {
        throw new Error('Neither OpenAI client nor recordings available')
      }

      // Check if the model wants to call functions
      if (completion.choices[0].finish_reason === 'tool_calls' && completion.choices[0].message.tool_calls) {
        const toolCalls = completion.choices[0].message.tool_calls

        // Add assistant message with tool calls to history
        state.conversation.messages.push({
          role: 'assistant',
          content: null,
          tool_calls: toolCalls.map(tc => ({
            id: tc.id,
            type: 'function' as const,
            function: {
              name: tc.function.name,
              arguments: tc.function.arguments
            }
          }))
        })

        // Execute each tool call
        for (const toolCall of toolCalls) {
          const functionName = toolCall.function.name
          const functionArgs = JSON.parse(toolCall.function.arguments)

          let functionResult: string
          if (functionName === 'GetWeatherAsync') {
            const location = functionArgs.location || 'unknown'
            functionResult = await getWeatherAsync(location)
          } else if (functionName === 'GetCurrentTime') {
            functionResult = getCurrentTime()
          } else {
            functionResult = 'Unknown function'
          }

          // Add function result to conversation history
          state.conversation.messages.push({
            role: 'tool',
            tool_call_id: toolCall.id,
            content: functionResult
          })
        }
      } else {
        // Model provided a final response
        response = completion.choices[0].message.content || ''
        state.conversation.messages.push({
          role: 'assistant',
          content: response
        })
        break
      }
    } catch (error) {
      console.error('Error during LLM call:', error)
      response = 'I apologize, but I encountered an error while processing your request.'
      break
    }
  }

  if (!response) {
    response = "I apologize, but I wasn't able to complete your request."
  }

  // Send all messages generated during this turn as separate activities
  for (let i = startingHistoryCount; i < state.conversation.messages.length; i++) {
    const chatMessage = state.conversation.messages[i]
    const agentMessage = convertChatMessageToAgentProtocol(chatMessage)

    // Create an activity with the Agent Protocol message in the Value field
    const activity: any = {
      type: 'message',
      text: '',
      value: agentMessage
    }
    await context.sendActivity(activity)
  }
})
